Add a feature flag for supporting partitioned cookies

sysrqb opened the pull request:
#### 2c6a2ee6af964d870ae3527d8e3f9cedeb684bc6
<pre>
Add a feature flag for supporting partitioned cookies
<a href="https://bugs.webkit.org/show_bug.cgi?id=279547">https://bugs.webkit.org/show_bug.cgi?id=279547</a>
<a href="https://rdar.apple.com/problem/135829408">rdar://problem/135829408</a>

Reviewed by Sihui Liu.

This patch adds a new preference and propagates preferences changes to the
Network process whenever a page is created or destroyed. This extends the
behavior we currently use for updating Blob URL partitioning state. This patch
doesn&apos;t introduce any changes in the current behavior.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setOptInCookiePartitioningEnabled):
* Source/WebCore/platform/network/NetworkStorageSession.h:
(WebCore::NetworkStorageSession::isOptInCookiePartitioningEnabled const):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setOptInCookiePartitioningEnabled const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setOptInCookiePartitioningEnabled):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::isOptInCookiePartitioningEnabled const):
(WebKit::WebsiteDataStore::propagateSettingUpdatesToNetworkProcess):
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::updateBlobRegistryPartitioningState): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: <a href="https://commits.webkit.org/284076@main">https://commits.webkit.org/284076@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/081711fb7cb9c88fbe2064dbb8fe40e6649a0c83

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68397 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72464 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19542 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70514 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55585 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19358 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54615 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13015 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71464 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43695 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59065 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35078 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40363 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16485 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17899 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61515 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62320 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16833 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74156 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67645 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12368 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16097 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62065 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12407 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59143 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62086 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10008 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3622 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89424 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10399 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43590 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15827 "Found 6 new JSC stress test failures: wasm.yaml/wasm/fuzz/export-function.js.default-wasm, wasm.yaml/wasm/fuzz/memory.js.wasm-eager-jettison, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-no-cjit, wasm.yaml/wasm/v8/many-parameters.js.wasm-bbq (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44664 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45858 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44406 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -98,7 +98,7 @@ static bool shouldBlockCookies(WebFrame* frame, const URL& firstPartyForCookies,
     }
 
-    return true;
+    return !shouldAttemptHandlingPartitionedCookies;
```
Why would this not return true? If client specifies to block all third-party cookies (i.e. `WebProcess::singleton().thirdPartyCookieBlockingMode() == ThirdPartyCookieBlockingMode::All` above), then we don't need to care whether the cookie would be put in partitioned storage or non-partitioned storage, right?

For the initial implementation, I think we don't need to change the check in web process side. We could let network process decide the storage by sending it "whether to handle partitioned attribute when processing cookies".

sysrqb replied:
I am interpreting `ThirdPartyCookieBlockingMode::All` as "unpartitioned first-party cookies in a third-party context", instead of "all cookies available in a third-party context".  But we can start with continuing to block all cookies in a third-party context, as you suggested.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1895,15 +1895,33 @@ bool WebsiteDataStore::isBlobRegistryPartitioningEnabled() const
 {
     auto enabled = isBlobRegistryPartitioningEnabled();
     if (m_isBlobRegistryPartitioningEnabled == enabled)
         return;
```
Should only return if both values are unchanged.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1895,15 +1895,33 @@ bool WebsiteDataStore::isBlobRegistryPartitioningEnabled() const
+    enabled = isOptInCookiePartitioningEnabled();
+    if (m_isOptInCookiePartitioningEnabled == enabled)
+        return;
+    if (RefPtr networkProcess = networkProcessIfExists()) {
```
Do we pass the value when creating network process and adding existing session to it?

sysrqb replied:
Good catch, I lost that change when I started working on another computer.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1895,15 +1895,33 @@ bool WebsiteDataStore::isBlobRegistryPartitioningEnabled() const
-void WebsiteDataStore::updateBlobRegistryPartitioningState()
+bool WebsiteDataStore::isOptInCookiePartitioningEnabled() const
+{
+    return WTF::anyOf(m_processes, [] (const WebProcessProxy& process) {
```
Why do we want a per-page setting if one page could affect all the other pages in the same session?

sysrqb replied:
This should only be used for testing, and applications like Safari only expose a single setting for all pages. The idea is that when the preference is enabled on any page, then we enable it in the NetworkSession. We only disable it in the NetworkSession when it is disabled on all pages.

szewai commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -519,6 +519,13 @@ void NetworkSession::setShouldSendPrivateTokenIPCForTesting(bool enabled)
+void NetworkSession::setOptInCookiePartitioningEnabled(bool enabled)
+{
+    RELEASE_LOG(Storage, "NetworkSession::setOptInCookiePartitioningEnabled as %" PUBLIC_LOG_STRING " for session %" PRIu64, enabled ? "enabled" : "disabled", m_sessionID.toUInt64());
+    if (auto* storageSession = networkStorageSession())
```
`CheckedPtr storageSession`

szewai reviewed (commented):
We don't need to include test if there is no behavior change; and the current tests are not clear about what the correct behavior is (for example, I would expect something like "FAIL: partitioned attribute is dropped in saved cookie" in the expected file).

I would suggest adding corresponding test when a new behavior is implemented.

szewai reviewed (commented):
The change looks fine to me, but I think we need to change the title and commit message to say "Add a feature flag..." instead of adding support after all the updates.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1895,15 +1895,29 @@ bool WebsiteDataStore::isBlobRegistryPartitioningEnabled() const
-    if (m_isBlobRegistryPartitioningEnabled == enabled)
-        return;
-    if (RefPtr networkProcess = networkProcessIfExists()) {
+    if (RefPtr networkProcess = m_isBlobRegistryPartitioningEnabled != enabled ? networkProcessIfExists() : nullptr) {
```
We could avoid duplicate check on network process by:
```
RefPtr networkProcess = networkProcessIfExists();
if (! networkProcess)
    return;

if (m_isBlobRegistryPartitioningEnabled != enabled) {
    ...
}

enabled = isOptInCookiePartitioningEnabled();
if (m_isOptInCookiePartitioningEnabled != enabled) {
    ...
}
```

szewai reviewed (approved):
LGTM with nits

Commit:
Add a feature flag for supporting partitioned cookies
https://bugs.webkit.org/show_bug.cgi?id=279547
rdar://problem/135829408

Reviewed by Sihui Liu.

This patch adds a new preference and propagates preferences changes to the
Network process whenever a page is created or destroyed. This extends the
behavior we currently use for updating Blob URL partitioning state. This patch
doesn't introduce any changes in the current behavior.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setOptInCookiePartitioningEnabled):
* Source/WebCore/platform/network/NetworkStorageSession.h:
(WebCore::NetworkStorageSession::isOptInCookiePartitioningEnabled const):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setOptInCookiePartitioningEnabled const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setOptInCookiePartitioningEnabled):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::isOptInCookiePartitioningEnabled const):
(WebKit::WebsiteDataStore::propagateSettingUpdatesToNetworkProcess):
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::updateBlobRegistryPartitioningState): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: https://commits.webkit.org/284076@main
